File: regions/system/library/dart/auth/test/auth_client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:k1s0_auth/auth.dart';
import 'package:test/test.dart';

final _testDiscovery = {
  'authorization_endpoint': 'https://auth.example.com/authorize',
  'token_endpoint': 'https://auth.example.com/token',
  'end_session_endpoint': 'https://auth.example.com/logout',
  'jwks_uri': 'https://auth.example.com/certs',
  'issuer': 'https://auth.example.com/realms/k1s0',
};

final _testConfig = AuthConfig(
  discoveryUrl: 'https://auth.example.com/.well-known/openid-configuration',
  clientId: 'test-client',
  redirectUri: 'https://app.example.com/callback',
  scopes: ['openid', 'profile', 'email'],
  postLogoutRedirectUri: 'https://app.example.com/',
);

Map<String, dynamic> _defaultTokenResponse() => {
      'access_token': 'mock-access-token',
      'refresh_token': 'mock-refresh-token',
      'id_token': 'mock-id-token',
      'expires_in': 900,
      'token_type': 'Bearer',
    };

void main() {
  late MemoryTokenStore tokenStore;
  late String? redirectedUrl;
  late List<String> httpGetUrls;
  late List<String> httpPostUrls;
  late Map<String, dynamic> Function() tokenResponseFactory;

  Future<http.Response> mockGet(Uri url, {Map<String, String>? headers}) async {
    httpGetUrls.add(url.toString());
    if (url.toString() == _testConfig.discoveryUrl) {
      return http.Response(jsonEncode(_testDiscovery), 200);
    }
    return http.Response('Not Found', 404);
  }

  Future<http.Response> mockPost(Uri url,
      {Map<String, String>? headers, Object? body}) async {
    httpPostUrls.add(url.toString());
    if (url.toString() == _testDiscovery['token_endpoint']) {
      return http.Response(jsonEncode(tokenResponseFactory()), 200);
    }
    return http.Response('Not Found', 404);
  }

  AuthClient createClient({
    Future<http.Response> Function(Uri, {Map<String, String>? headers})?
        httpGet,
    Future<http.Response> Function(Uri,
            {Map<String, String>? headers, Object? body})?
        httpPost,
  }) {
    return AuthClient(AuthClientOptions(
      config: _testConfig,
      tokenStore: tokenStore,
      httpGet: httpGet ?? mockGet,
      httpPost: httpPost ?? mockPost,
      redirect: (url) => redirectedUrl = url,
      generateState: () => 'mock-state-value',
    ));
  }

  setUp(() {
    tokenStore = MemoryTokenStore();
    redirectedUrl = null;
    httpGetUrls = [];
    httpPostUrls = [];
    tokenResponseFactory = _defaultTokenResponse;
  });

  group('AuthClient', () {
    group('login', () {
      test('PKCE パラメータ付きで認可エンドポイントにリダイレクトすること',
          () async {
        final client = createClient();
        await client.login();

        expect(redirectedUrl, isNotNull);
        final uri = Uri.parse(redirectedUrl!);
        expect(
          '${uri.scheme}://${uri.host}${uri.path}',
          equals(_testDiscovery['authorization_endpoint']),
        );
        expect(uri.queryParameters['response_type'], equals('code'));
        expect(uri.queryParameters['client_id'], equals('test-client'));
        expect(uri.queryParameters['redirect_uri'],
            equals('https://app.example.com/callback'));
        expect(uri.queryParameters['scope'], equals('openid profile email'));
        expect(uri.queryParameters['code_challenge_method'], equals('S256'));
        expect(uri.queryParameters['code_challenge'], isNotEmpty);
        expect(uri.queryParameters['state'], equals('mock-state-value'));
      });

      test('code_verifier と state を保存すること', () async {
        final client = createClient();
        await client.login();

        expect(tokenStore.getCodeVerifier(), isNotNull);
        expect(tokenStore.getCodeVerifier(), isNotEmpty);
        expect(tokenStore.getState(), equals('mock-state-value'));
      });

      test('OIDC ディスカバリドキュメントを取得すること', () async {
        final client = createClient();
        await client.login();

        expect(httpGetUrls, contains(_testConfig.discoveryUrl));
      });
    });

    group('handleCallback', () {
      test('コードをトークンに交換すること', () async {
        final client = createClient();
        await client.login();

        final tokenSet =
            await client.handleCallback('auth-code-123', 'mock-state-value');

        expect(tokenSet.accessToken, equals('mock-access-token'));
        expect(tokenSet.refreshToken, equals('mock-refresh-token'));
        expect(tokenSet.idToken, equals('mock-id-token'));
        expect(tokenSet.isValid, isTrue);
      });

      test('state が一致しない場合に例外をスローすること', () async {
        final client = createClient();
        await client.login();

        expect(
          () => client.handleCallback('code', 'wrong-state'),
          throwsA(isA<AuthError>().having(
            (e) => e.message,
            'message',
            'State mismatch',
          )),
        );
      });

      test('PKCE ベリファイアが存在しない場合に例外をスローすること', () async {
        final client = createClient();
        tokenStore.setState('mock-state-value');
        // Don't set code verifier

        expect(
          () => client.handleCallback('code', 'mock-state-value'),
          throwsA(isA<AuthError>().having(
            (e) => e.message,
            'message',
            'Missing PKCE verifier',
          )),
        );
      });

      test('トークンリクエストが失敗した場合に例外をスローすること', () async {
        Future<http.Response> failPost(Uri url,
            {Map<String, String>? headers, Object? body}) async {
          httpPostUrls.add(url.toString());
          return http.Response('Unauthorized', 401);
        }

        final client = createClient(httpPost: failPost);
        await client.login();

        expect(
          () => client.handleCallback('code', 'mock-state-value'),
          throwsA(isA<AuthError>().having(
            (e) => e.message,
            'message',
            contains('Token request failed'),
          )),
        );
      });

      test('コールバック成功時にリスナーに通知すること', () async {
        final client = createClient();
        final states = <bool>[];
        client.onAuthStateChange(states.add);

        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        expect(states, contains(true));
      });

      test('コールバック成功後に code verifier と state をクリアすること',
          () async {
        final client = createClient();
        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        expect(tokenStore.getCodeVerifier(), isNull);
        expect(tokenStore.getState(), isNull);
      });

      test('トークンセットを保存すること', () async {
        final client = createClient();
        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        final stored = tokenStore.getTokenSet();
        expect(stored, isNotNull);
        expect(stored!.accessToken, equals('mock-access-token'));
      });
    });

    group('getAccessToken', () {
      test('有効なアクセストークンを返すこと', () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'valid-token',
          refreshToken: 'refresh-token',
          idToken: 'id-token',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));

        final client = createClient();
        final token = await client.getAccessToken();
        expect(token, equals('valid-token'));
      });

      test('未認証の場合に例外をスローすること', () async {
        final client = createClient();
        expect(
          client.getAccessToken,
          throwsA(isA<AuthError>().having(
            (e) => e.message,
            'message',
            'Not authenticated',
          )),
        );
      });

      test('トークンの有効期限が 60 秒以内の場合に自動リフレッシュすること',
          () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'expiring-token',
          refreshToken: 'refresh-token',
          idToken: 'id-token',
          expiresAt: DateTime.now().add(const Duration(seconds: 30)),
        ));

        tokenResponseFactory = () => {
              'access_token': 'refreshed-token',
              'refresh_token': 'new-refresh-token',
              'id_token': 'new-id-token',
              'expires_in': 900,
              'token_type': 'Bearer',
            };

        final client = createClient();
        final token = await client.getAccessToken();
        expect(token, equals('refreshed-token'));
      });
    });

    group('refreshToken', () {
      test('リフレッシュトークンを新しいトークンと交換すること', () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'old-access',
          refreshToken: 'old-refresh',
          idToken: 'old-id',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));

        final client = createClient();
        await client.refreshToken();

        final newTokenSet = tokenStore.getTokenSet();
        expect(newTokenSet!.accessToken, equals('mock-access-token'));
        expect(newTokenSet.refreshToken, equals('mock-refresh-token'));
      });

      test('リフレッシュトークンが存在しない場合に例外をスローすること', () async {
        final client = createClient();
        expect(
          client.refreshToken,
          throwsA(isA<AuthError>()),
        );
      });

      test('リフレッシュ失敗時にトークンをクリアしてリスナーに通知すること',
          () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'access',
          refreshToken: 'expired-refresh',
          idToken: 'id',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));

        Future<http.Response> failPost(Uri url,
            {Map<String, String>? headers, Object? body}) async {
          httpPostUrls.add(url.toString());
          return http.Response('Forbidden', 403);
        }

        final client = createClient(httpPost: failPost);
        final states = <bool>[];
        client.onAuthStateChange(states.add);

        expect(
          client.refreshToken,
          throwsA(isA<AuthError>().having(
            (e) => e.message,
            'message',
            contains('Token refresh failed'),
          )),
        );
      });
    });

    group('isAuthenticated', () {
      test('トークンセットが存在しない場合に false を返すこと', () {
        final client = createClient();
        expect(client.isAuthenticated, isFalse);
      });

      test('トークンが有効な場合に true を返すこと', () {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'token',
          refreshToken: 'refresh',
          idToken: 'id',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));
        final client = createClient();
        expect(client.isAuthenticated, isTrue);
      });

      test('トークンが期限切れの場合に false を返すこと', () {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'token',
          refreshToken: 'refresh',
          idToken: 'id',
          expiresAt: DateTime.now().subtract(const Duration(seconds: 1)),
        ));
        final client = createClient();
        expect(client.isAuthenticated, isFalse);
      });
    });

    group('logout', () {
      test('トークンをクリアすること', () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'token',
          refreshToken: 'refresh',
          idToken: 'id',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));
        final client = createClient();
        await client.logout();

        expect(tokenStore.getTokenSet(), isNull);
      });

      test('リスナーに通知すること', () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'token',
          refreshToken: 'refresh',
          idToken: 'id',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));
        final client = createClient();
        final states = <bool>[];
        client.onAuthStateChange(states.add);
        await client.logout();

        expect(states, contains(false));
      });

      test(
          'id_token_hint 付きで end_session_endpoint にリダイレクトすること',
          () async {
        tokenStore.setTokenSet(TokenSet(
          accessToken: 'token',
          refreshToken: 'refresh',
          idToken: 'my-id-token',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        ));
        final client = createClient();
        await client.logout();

        expect(redirectedUrl, isNotNull);
        final uri = Uri.parse(redirectedUrl!);
        expect(
          '${uri.scheme}://${uri.host}${uri.path}',
          equals(_testDiscovery['end_session_endpoint']),
        );
        expect(uri.queryParameters['id_token_hint'], equals('my-id-token'));
        expect(
          uri.queryParameters['post_logout_redirect_uri'],
          equals(_testConfig.postLogoutRedirectUri),
        );
        expect(
          uri.queryParameters['client_id'],
          equals(_testConfig.clientId),
        );
      });

      test('トークンセットが存在しない場合はリダイレクトしないこと', () async {
        final client = createClient();
        await client.logout();

        expect(redirectedUrl, isNull);
      });
    });

    group('onAuthStateChange', () {
      test('リスナーを登録して通知すること', () async {
        final client = createClient();
        final states = <bool>[];
        client.onAuthStateChange(states.add);

        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        expect(states, contains(true));
      });

      test('購読解除関数を返すこと', () async {
        final client = createClient();
        final states = <bool>[];
        final unsubscribe =
            client.onAuthStateChange(states.add);

        unsubscribe();

        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        expect(states, isEmpty);
      });

      test('複数のリスナーをサポートすること', () async {
        final client = createClient();
        final states1 = <bool>[];
        final states2 = <bool>[];
        client.onAuthStateChange(states1.add);
        client.onAuthStateChange(states2.add);

        await client.login();
        await client.handleCallback('code', 'mock-state-value');

        expect(states1, contains(true));
        expect(states2, contains(true));
      });
    });

    group('getTokenSet', () {
      test('トークンが存在しない場合に null を返すこと', () {
        final client = createClient();
        expect(client.getTokenSet(), isNull);
      });

      test('保存されたトークンセットを返すこと', () {
        final ts = TokenSet(
          accessToken: 'a',
          refreshToken: 'r',
          idToken: 'i',
          expiresAt: DateTime.now().add(const Duration(minutes: 5)),
        );
        tokenStore.setTokenSet(ts);
        final client = createClient();
        final result = client.getTokenSet();
        expect(result, isNotNull);
        expect(result!.accessToken, equals('a'));
      });
    });

    group('getAuthorizationUrl', () {
      test('リダイレクトせずに認可 URL を返すこと', () async {
        final client = createClient();
        final url = await client.getAuthorizationUrl();

        expect(redirectedUrl, isNull); // should NOT redirect
        final uri = Uri.parse(url);
        expect(
          '${uri.scheme}://${uri.host}${uri.path}',
          equals(_testDiscovery['authorization_endpoint']),
        );
        expect(uri.queryParameters['response_type'], equals('code'));
        expect(uri.queryParameters['code_challenge_method'], equals('S256'));
      });
    });

    group('discovery caching', () {
      test('ディスカバリドキュメントをキャッシュすること', () async {
        final client = createClient();
        await client.login();
        redirectedUrl = null;
        await client.login();

        final discoveryCalls = httpGetUrls
            .where((url) => url == _testConfig.discoveryUrl)
            .length;
        expect(discoveryCalls, equals(1));
      });
    });

    group('AuthError', () {
      test('正しいメッセージを持つこと', () {
        final error = AuthError('test error');
        expect(error.message, equals('test error'));
        expect(error.toString(), equals('AuthError: test error'));
      });
    });
  });

  group('TokenSet', () {
    test('有効期限が未来の場合に isValid が true を返すこと', () {
      final ts = TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now().add(const Duration(minutes: 5)),
      );
      expect(ts.isValid, isTrue);
    });

    test('有効期限が過去の場合に isValid が false を返すこと', () {
      final ts = TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now().subtract(const Duration(seconds: 1)),
      );
      expect(ts.isValid, isFalse);
    });

    test('閾値内の場合に isExpiringSoon が true を返すこと', () {
      final ts = TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now().add(const Duration(seconds: 30)),
      );
      expect(ts.isExpiringSoon(), isTrue);
    });

    test('閾値外の場合に isExpiringSoon が false を返すこと', () {
      final ts = TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now().add(const Duration(minutes: 5)),
      );
      expect(ts.isExpiringSoon(), isFalse);
    });

    test('toJson と fromJson でラウンドトリップできること', () {
      final ts = TokenSet(
        accessToken: 'access',
        refreshToken: 'refresh',
        idToken: 'id',
        expiresAt: DateTime.utc(2026, 1, 1),
      );
      final json = ts.toJson();
      final restored = TokenSet.fromJson(json);
      expect(restored.accessToken, equals('access'));
      expect(restored.refreshToken, equals('refresh'));
      expect(restored.idToken, equals('id'));
      expect(restored.expiresAt, equals(DateTime.utc(2026, 1, 1)));
    });
  });

  group('MemoryTokenStore', () {
    test('トークンセットを保存・取得できること', () {
      final store = MemoryTokenStore();
      final ts = TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now().add(const Duration(minutes: 5)),
      );
      store.setTokenSet(ts);
      expect(store.getTokenSet()?.accessToken, equals('a'));
    });

    test('トークンセットをクリアできること', () {
      final store = MemoryTokenStore();
      store.setTokenSet(TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now(),
      ));
      store.clearTokenSet();
      expect(store.getTokenSet(), isNull);
    });

    test('code verifier を保存・取得できること', () {
      final store = MemoryTokenStore();
      store.setCodeVerifier('test-verifier');
      expect(store.getCodeVerifier(), equals('test-verifier'));
    });

    test('state を保存・取得できること', () {
      final store = MemoryTokenStore();
      store.setState('test-state');
      expect(store.getState(), equals('test-state'));
    });

    test('clearAll で全データをクリアすること', () {
      final store = MemoryTokenStore();
      store.setTokenSet(TokenSet(
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: DateTime.now(),
      ));
      store.setCodeVerifier('verifier');
      store.setState('state');
      store.clearAll();
      expect(store.getTokenSet(), isNull);
      expect(store.getCodeVerifier(), isNull);
      expect(store.getState(), isNull);
    });
  });
}

File: regions/system/library/typescript/auth/__tests__/auth-client.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { AuthClient, AuthError } from '../src/auth-client';
import { MemoryTokenStore } from '../src/token-store';
import type { AuthConfig, TokenSet, OIDCDiscovery } from '../src/types';

const TEST_DISCOVERY: OIDCDiscovery = {
  authorization_endpoint: 'https://auth.example.com/authorize',
  token_endpoint: 'https://auth.example.com/token',
  end_session_endpoint: 'https://auth.example.com/logout',
  jwks_uri: 'https://auth.example.com/certs',
  issuer: 'https://auth.example.com/realms/k1s0',
};

const TEST_CONFIG: AuthConfig = {
  discoveryUrl: 'https://auth.example.com/.well-known/openid-configuration',
  clientId: 'test-client',
  redirectUri: 'https://app.example.com/callback',
  scopes: ['openid', 'profile', 'email'],
  postLogoutRedirectUri: 'https://app.example.com/',
};

function createMockFetch(responses: Record<string, unknown> = {}) {
  return vi.fn(async (input: RequestInfo | URL, _init?: RequestInit) => {
    const url = typeof input === 'string' ? input : input.toString();

    if (url === TEST_CONFIG.discoveryUrl) {
      return new Response(JSON.stringify(TEST_DISCOVERY), { status: 200 });
    }

    if (url === TEST_DISCOVERY.token_endpoint) {
      const tokenResponse = responses['token'] ?? {
        access_token: 'mock-access-token',
        refresh_token: 'mock-refresh-token',
        id_token: 'mock-id-token',
        expires_in: 900,
        token_type: 'Bearer',
      };
      return new Response(JSON.stringify(tokenResponse), { status: 200 });
    }

    return new Response('Not Found', { status: 404 });
  });
}

describe('AuthClient', () => {
  let tokenStore: MemoryTokenStore;
  let redirectUrl: string | null;
  let mockFetch: ReturnType<typeof createMockFetch>;

  beforeEach(() => {
    tokenStore = new MemoryTokenStore();
    redirectUrl = null;
    mockFetch = createMockFetch();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  function createClient(overrides: Partial<ConstructorParameters<typeof AuthClient>[0]> = {}) {
    return new AuthClient({
      config: TEST_CONFIG,
      tokenStore,
      fetch: mockFetch,
      redirect: (url: string) => {
        redirectUrl = url;
      },
      generateState: () => 'mock-state-value',
      ...overrides,
    });
  }

  describe('login', () => {
    it('should redirect to the authorization endpoint with PKCE params', async () => {
      const client = createClient();
      await client.login();

      expect(redirectUrl).not.toBeNull();
      const url = new URL(redirectUrl!);
      expect(url.origin + url.pathname).toBe(TEST_DISCOVERY.authorization_endpoint);
      expect(url.searchParams.get('response_type')).toBe('code');
      expect(url.searchParams.get('client_id')).toBe(TEST_CONFIG.clientId);
      expect(url.searchParams.get('redirect_uri')).toBe(TEST_CONFIG.redirectUri);
      expect(url.searchParams.get('scope')).toBe('openid profile email');
      expect(url.searchParams.get('code_challenge_method')).toBe('S256');
      expect(url.searchParams.get('code_challenge')).toBeTruthy();
      expect(url.searchParams.get('state')).toBe('mock-state-value');
    });

    it('should store code_verifier and state', async () => {
      const client = createClient();
      await client.login();

      expect(tokenStore.getCodeVerifier()).toBeTruthy();
      expect(tokenStore.getState()).toBe('mock-state-value');
    });

    it('should fetch the OIDC discovery document', async () => {
      const client = createClient();
      await client.login();

      expect(mockFetch).toHaveBeenCalledWith(TEST_CONFIG.discoveryUrl);
    });
  });

  describe('handleCallback', () => {
    it('should exchange code for tokens', async () => {
      const client = createClient();
      // Setup: call login first to store verifier and state
      await client.login();

      vi.setSystemTime(new Date('2026-01-01T00:00:00Z'));
      const tokenSet = await client.handleCallback('auth-code-123', 'mock-state-value');

      expect(tokenSet.accessToken).toBe('mock-access-token');
      expect(tokenSet.refreshToken).toBe('mock-refresh-token');
      expect(tokenSet.idToken).toBe('mock-id-token');
      expect(tokenSet.expiresAt).toBe(Date.now() + 900_000);
    });

    it('should send correct parameters to token endpoint', async () => {
      const client = createClient();
      await client.login();
      await client.handleCallback('auth-code-123', 'mock-state-value');

      // Find the token endpoint call
      const tokenCall = mockFetch.mock.calls.find(
        ([url]) => url === TEST_DISCOVERY.token_endpoint,
      );
      expect(tokenCall).toBeTruthy();

      const body = tokenCall![1]?.body as URLSearchParams;
      expect(body.get('grant_type')).toBe('authorization_code');
      expect(body.get('client_id')).toBe(TEST_CONFIG.clientId);
      expect(body.get('code')).toBe('auth-code-123');
      expect(body.get('redirect_uri')).toBe(TEST_CONFIG.redirectUri);
      expect(body.get('code_verifier')).toBeTruthy();
    });

    it('should throw on state mismatch', async () => {
      const client = createClient();
      await client.login();

      await expect(
        client.handleCallback('code', 'wrong-state'),
      ).rejects.toThrow('State mismatch');
    });

    it('should throw when PKCE verifier is missing', async () => {
      const client = createClient();
      tokenStore.setState('mock-state-value');
      // Don't set code verifier

      await expect(
        client.handleCallback('code', 'mock-state-value'),
      ).rejects.toThrow('Missing PKCE verifier');
    });

    it('should throw on token request failure', async () => {
      const failFetch = vi.fn(async (input: RequestInfo | URL) => {
        const url = typeof input === 'string' ? input : input.toString();
        if (url === TEST_CONFIG.discoveryUrl) {
          return new Response(JSON.stringify(TEST_DISCOVERY), { status: 200 });
        }
        return new Response('Unauthorized', { status: 401 });
      });

      const client = createClient({ fetch: failFetch });
      await client.login();

      await expect(
        client.handleCallback('code', 'mock-state-value'),
      ).rejects.toThrow('Token request failed: 401');
    });

    it('should notify listeners on successful callback', async () => {
      const client = createClient();
      const listener = vi.fn();
      client.onAuthStateChange(listener);

      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      expect(listener).toHaveBeenCalledWith(true);
    });

    it('should clear code verifier and state after successful callback', async () => {
      const client = createClient();
      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      expect(tokenStore.getCodeVerifier()).toBeNull();
      expect(tokenStore.getState()).toBeNull();
    });

    it('should store the token set', async () => {
      const client = createClient();
      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      const stored = tokenStore.getTokenSet();
      expect(stored).not.toBeNull();
      expect(stored!.accessToken).toBe('mock-access-token');
    });
  });

  describe('getAccessToken', () => {
    it('should return the access token when valid', async () => {
      vi.setSystemTime(new Date('2026-01-01T00:00:00Z'));
      tokenStore.setTokenSet({
        accessToken: 'valid-token',
        refreshToken: 'refresh-token',
        idToken: 'id-token',
        expiresAt: Date.now() + 300_000, // 5 minutes from now
      });

      const client = createClient();
      const token = await client.getAccessToken();
      expect(token).toBe('valid-token');
    });

    it('should throw when not authenticated', async () => {
      const client = createClient();
      await expect(client.getAccessToken()).rejects.toThrow('Not authenticated');
    });

    it('should auto-refresh when token expires within 60 seconds', async () => {
      vi.setSystemTime(new Date('2026-01-01T00:00:00Z'));
      tokenStore.setTokenSet({
        accessToken: 'expiring-token',
        refreshToken: 'refresh-token',
        idToken: 'id-token',
        expiresAt: Date.now() + 30_000, // 30 seconds from now (< 60s threshold)
      });

      const refreshedFetch = createMockFetch({
        token: {
          access_token: 'refreshed-token',
          refresh_token: 'new-refresh-token',
          id_token: 'new-id-token',
          expires_in: 900,
          token_type: 'Bearer',
        },
      });

      const client = createClient({ fetch: refreshedFetch });
      const token = await client.getAccessToken();
      expect(token).toBe('refreshed-token');
    });
  });

  describe('refreshToken', () => {
    it('should exchange refresh token for new tokens', async () => {
      vi.setSystemTime(new Date('2026-01-01T00:00:00Z'));
      tokenStore.setTokenSet({
        accessToken: 'old-access',
        refreshToken: 'old-refresh',
        idToken: 'old-id',
        expiresAt: Date.now() + 100_000,
      });

      const client = createClient();
      await client.refreshToken();

      const newTokenSet = tokenStore.getTokenSet();
      expect(newTokenSet!.accessToken).toBe('mock-access-token');
      expect(newTokenSet!.refreshToken).toBe('mock-refresh-token');
    });

    it('should throw when no refresh token is available', async () => {
      const client = createClient();
      await expect(client.refreshToken()).rejects.toThrow('No refresh token');
    });

    it('should send correct refresh parameters', async () => {
      tokenStore.setTokenSet({
        accessToken: 'access',
        refreshToken: 'my-refresh-token',
        idToken: 'id',
        expiresAt: Date.now() + 100_000,
      });

      const client = createClient();
      await client.refreshToken();

      const tokenCall = mockFetch.mock.calls.find(
        ([url]) => url === TEST_DISCOVERY.token_endpoint,
      );
      const body = tokenCall![1]?.body as URLSearchParams;
      expect(body.get('grant_type')).toBe('refresh_token');
      expect(body.get('client_id')).toBe(TEST_CONFIG.clientId);
      expect(body.get('refresh_token')).toBe('my-refresh-token');
    });

    it('should clear tokens and notify listeners on refresh failure', async () => {
      tokenStore.setTokenSet({
        accessToken: 'access',
        refreshToken: 'expired-refresh',
        idToken: 'id',
        expiresAt: Date.now() + 100_000,
      });

      const failFetch = vi.fn(async (input: RequestInfo | URL) => {
        const url = typeof input === 'string' ? input : input.toString();
        if (url === TEST_CONFIG.discoveryUrl) {
          return new Response(JSON.stringify(TEST_DISCOVERY), { status: 200 });
        }
        return new Response('Forbidden', { status: 403 });
      });

      const client = createClient({ fetch: failFetch });
      const listener = vi.fn();
      client.onAuthStateChange(listener);

      await expect(client.refreshToken()).rejects.toThrow('Token refresh failed: 403');
      expect(tokenStore.getTokenSet()).toBeNull();
      expect(listener).toHaveBeenCalledWith(false);
    });
  });

  describe('isAuthenticated', () => {
    it('should return false when no token set', () => {
      const client = createClient();
      expect(client.isAuthenticated()).toBe(false);
    });

    it('should return true when token is valid', () => {
      tokenStore.setTokenSet({
        accessToken: 'token',
        refreshToken: 'refresh',
        idToken: 'id',
        expiresAt: Date.now() + 300_000,
      });
      const client = createClient();
      expect(client.isAuthenticated()).toBe(true);
    });

    it('should return false when token has expired', () => {
      tokenStore.setTokenSet({
        accessToken: 'token',
        refreshToken: 'refresh',
        idToken: 'id',
        expiresAt: Date.now() - 1000,
      });
      const client = createClient();
      expect(client.isAuthenticated()).toBe(false);
    });
  });

  describe('logout', () => {
    it('should clear tokens', async () => {
      tokenStore.setTokenSet({
        accessToken: 'token',
        refreshToken: 'refresh',
        idToken: 'id',
        expiresAt: Date.now() + 300_000,
      });
      const client = createClient();
      await client.logout();

      expect(tokenStore.getTokenSet()).toBeNull();
    });

    it('should notify listeners', async () => {
      tokenStore.setTokenSet({
        accessToken: 'token',
        refreshToken: 'refresh',
        idToken: 'id',
        expiresAt: Date.now() + 300_000,
      });
      const client = createClient();
      const listener = vi.fn();
      client.onAuthStateChange(listener);
      await client.logout();

      expect(listener).toHaveBeenCalledWith(false);
    });

    it('should redirect to end_session_endpoint with id_token_hint', async () => {
      tokenStore.setTokenSet({
        accessToken: 'token',
        refreshToken: 'refresh',
        idToken: 'my-id-token',
        expiresAt: Date.now() + 300_000,
      });
      const client = createClient();
      await client.logout();

      expect(redirectUrl).not.toBeNull();
      const url = new URL(redirectUrl!);
      expect(url.origin + url.pathname).toBe(TEST_DISCOVERY.end_session_endpoint);
      expect(url.searchParams.get('id_token_hint')).toBe('my-id-token');
      expect(url.searchParams.get('post_logout_redirect_uri')).toBe(TEST_CONFIG.postLogoutRedirectUri);
      expect(url.searchParams.get('client_id')).toBe(TEST_CONFIG.clientId);
    });

    it('should not redirect when no token set exists', async () => {
      const client = createClient();
      await client.logout();

      expect(redirectUrl).toBeNull();
    });
  });

  describe('onAuthStateChange', () => {
    it('should register and notify a listener', async () => {
      const client = createClient();
      const listener = vi.fn();
      client.onAuthStateChange(listener);

      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      expect(listener).toHaveBeenCalledWith(true);
    });

    it('should return an unsubscribe function', async () => {
      const client = createClient();
      const listener = vi.fn();
      const unsubscribe = client.onAuthStateChange(listener);

      unsubscribe();

      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      expect(listener).not.toHaveBeenCalled();
    });

    it('should support multiple listeners', async () => {
      const client = createClient();
      const listener1 = vi.fn();
      const listener2 = vi.fn();
      client.onAuthStateChange(listener1);
      client.onAuthStateChange(listener2);

      await client.login();
      await client.handleCallback('code', 'mock-state-value');

      expect(listener1).toHaveBeenCalledWith(true);
      expect(listener2).toHaveBeenCalledWith(true);
    });
  });

  describe('getTokenSet', () => {
    it('should return null when no tokens', () => {
      const client = createClient();
      expect(client.getTokenSet()).toBeNull();
    });

    it('should return stored token set', () => {
      const ts: TokenSet = {
        accessToken: 'a',
        refreshToken: 'r',
        idToken: 'i',
        expiresAt: Date.now() + 300_000,
      };
      tokenStore.setTokenSet(ts);
      const client = createClient();
      expect(client.getTokenSet()).toEqual(ts);
    });
  });

  describe('discovery caching', () => {
    it('should cache the discovery document', async () => {
      const client = createClient();
      await client.login();
      // Reset the redirect to detect a second login redirect
      redirectUrl = null;
      await client.login();

      // discovery endpoint should only be fetched once (cached on second call)
      const discoveryCalls = mockFetch.mock.calls.filter(
        ([url]) => url === TEST_CONFIG.discoveryUrl,
      );
      expect(discoveryCalls.length).toBe(1);
    });
  });

  describe('AuthError', () => {
    it('should have the correct name and message', () => {
      const error = new AuthError('test error');
      expect(error.name).toBe('AuthError');
      expect(error.message).toBe('test error');
      expect(error).toBeInstanceOf(Error);
    });
  });
});
